### ISSUE TITLE
py-basemap (and py-matplotlib too): Doesn't work

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#1948]
@adamjstewart 

I was able to build `py-basemap` OK, but when I try to use it I get:

```
>>> import mpl_toolkits.basemap
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home2/rpfische/spack3/opt/spack/linux-centos7-x86_64/gcc-4.9.3/py-basemap-1.0.7-ruhltkr62akc3yp2llvsqupikvhkq5b5/lib/python3.5/site-packages/mpl_toolkits/basemap/__init__.py", line 31, in <module>
    from mpl_toolkits.axes_grid1 import make_axes_locatable
ImportError: No module named 'mpl_toolkits.axes_grid1'
```

Maybe this is related to my issue: http://stackoverflow.com/questions/16769899/matplotlib-basemap-no-module-named-axes-grid1

Note that I am using Python3.  Here's the entire spec:

```
  py-basemap@1.0.7%gcc@4.9.3 arch=linux-centos7-x86_64
      ^geos@3.5.0%gcc@4.9.3 arch=linux-centos7-x86_64
      ^py-matplotlib@1.5.1%gcc@4.9.3+gui+ipython arch=linux-centos7-x86_64
          ^ImageMagick@7.0.2-7%gcc@4.9.3 arch=linux-centos7-x86_64
              ^fontconfig@2.11.1%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^freetype@2.5.3%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^libpng@1.6.24%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^zlib@1.2.8%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^libxml2@2.9.4%gcc@4.9.3~python arch=linux-centos7-x86_64
                      ^xz@5.2.2%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^pkg-config@0.29.1%gcc@4.9.3 arch=linux-centos7-x86_64
              ^ghostscript@9.18%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^libtiff@4.0.6%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^jpeg@9b%gcc@4.9.3 arch=linux-centos7-x86_64
              ^ghostscript-fonts@8.11%gcc@4.9.3 arch=linux-centos7-x86_64
              ^libtool@2.4.6%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^m4@1.4.17%gcc@4.9.3+sigsegv arch=linux-centos7-x86_64
                      ^libsigsegv@2.10%gcc@4.9.3 arch=linux-centos7-x86_64
              ^pango@1.40.1%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^cairo@1.14.0%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^glib@2.49.4%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^autoconf@2.69%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^automake@1.15%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^gettext@0.19.8.1%gcc@4.9.3+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-centos7-x86_64
                              ^bzip2@1.0.6%gcc@4.9.3 arch=linux-centos7-x86_64
                              ^ncurses@6.0%gcc@4.9.3 arch=linux-centos7-x86_64
                              ^tar@1.29%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^libffi@3.2.1%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^pcre@8.39%gcc@4.9.3+utf arch=linux-centos7-x86_64
                      ^pixman@0.32.6%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^harfbuzz@0.9.37%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^icu4c@57.1%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-cycler@0.10.0%gcc@4.9.3 arch=linux-centos7-x86_64
              ^py-setuptools@25.2.0%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^python@3.5.2%gcc@4.9.3~tk~ucs4 arch=linux-centos7-x86_64
                      ^openssl@system%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^readline@6.3%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^sqlite@3.8.5%gcc@4.9.3 arch=linux-centos7-x86_64
              ^py-six@1.10.0%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-dateutil@2.5.2%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-ipython@3.1.0%gcc@4.9.3 arch=linux-centos7-x86_64
              ^py-pygments@2.1.3%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-mock@1.3.0%gcc@4.9.3 arch=linux-centos7-x86_64
              ^py-pbr@1.8.1%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-nose@1.3.7%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-numpy@1.11.1%gcc@4.9.3+blas+lapack arch=linux-centos7-x86_64
              ^openblas@0.2.19%gcc@4.9.3+fpic~openmp+shared arch=linux-centos7-x86_64
          ^py-pillow@3.2.0%gcc@4.9.3~freetype+jpeg~jpeg2000~lcms~tiff+zlib arch=linux-centos7-x86_64
              ^binutils@2.27%gcc@4.9.3+gold~krellpatch~libiberty arch=linux-centos7-x86_64
                  ^bison@3.0.4%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^flex@2.6.0%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-pyparsing@2.0.3%gcc@4.9.3 arch=linux-centos7-x86_64
          ^py-pyside@1.2.4%gcc@4.9.3 arch=linux-centos7-x86_64
              ^cmake@3.6.1%gcc@4.9.3~doc+ncurses+openssl~ownlibs~qt arch=linux-centos7-x86_64
                  ^curl@7.50.3%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^expat@2.2.0%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^libarchive@3.2.1%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^lz4@131%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^lzma@4.32.7%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^lzo@2.09%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^nettle@3.2%gcc@4.9.3 arch=linux-centos7-x86_64
                          ^gmp@6.1.1%gcc@4.9.3 arch=linux-centos7-x86_64
              ^libxslt@1.1.28%gcc@4.9.3 arch=linux-centos7-x86_64
                  ^libgcrypt@1.6.2%gcc@4.9.3 arch=linux-centos7-x86_64
                      ^libgpg-error@1.21%gcc@4.9.3 arch=linux-centos7-x86_64
              ^qt@4.8.5%gcc@4.9.3~dbus~examples~gtk~krellpatch~mesa arch=linux-centos7-x86_64
          ^py-pytz@2016.6.1%gcc@4.9.3 arch=linux-centos7-x86_64
          ^qhull@2015.2%gcc@4.9.3 arch=linux-centos7-x86_64
          ^tk@8.6.5%gcc@4.9.3 arch=linux-centos7-x86_64
              ^tcl@8.6.5%gcc@4.9.3 arch=linux-centos7-x86_64
```

BTW... are all these dependencies really necessary?  I notice a lot have been added, and I'm wondering what they've brought me (other than extra build time).  Does the default configuration have them mostly turned off or mostly on?

**Comment by citibeth** [comment]
Among other things... I think that throwing in the "kitchen sink" GUI is probably not what we want.  Better would be '+tk' and '+qt' variants.

If ImageMagick is optional, why is there not a variant for it?  Do we need ImageMagick and PIL, or is just one or the other sufficient?  The ImageMagick build seems to be a real hog.

Can the testing dependencies be implemented as type='build' dependencies?

**Comment by adamjstewart** [comment]
From the [Matplotlib Installation Instructions](http://matplotlib.org/users/installing.html):

> Matplotlib depends on Pillow for reading and saving JPEG, BMP, and TIFF image files. Matplotlib requires MiKTeX and GhostScript for rendering text with LaTeX. FFmpeg, avconv, mencoder, or ImageMagick are required for the animation module.

Also see http://matplotlib.org/users/installing.html#optional-external-programs for more details.

So it sounds like both are optional, but they definitely cover different things.

I agree with respect to the GUI backend stuff. I didn't add any of those because it was way over my head. Hopefully someone else can figure out what is actually necessary for that.

The testing dependencies should be their own deptype. See #1279 for discussion.

**Comment by citibeth** [comment]
So, I would suggest the following variants:

* `+latex`: Adds MiKTeX and GhostScript.  (Does Spack do latex, or do we depend on the system?)
* `+animation`: Adds ImageMagick.
[If it works with ImageMagick, will people want to do the other animation options as well?]

> I didn't add any of those because it was way over my head. Hopefully someone else can figure out what is actually necessary for that.

Can you explain further?  Why not just:
```
    depends_on('tk@8.3:',   when='+tk')  # not 8.6.0 or 8.6.1
    depends_on('qt',        when='+qt')
    depends_on('py-pyside', when='+qt', type=nolink)
```
...plus maybe a bit of change on the filtering in the backends file at the end of the package?

> The testing dependencies should be their own deptype. See #1279 for discussion.

In the meantime, can we  use `type='build'` for them?

**Comment by adamjstewart** [comment]
> So, I would suggest the following variants:

Works for me. Spack has `texinfo` and `texlive` packages, which I believe are the equivalent of `MikTeX` for Linux.

> Why not just:
```
    depends_on('tk@8.3:',   when='+tk')  # not 8.6.0 or 8.6.1
    depends_on('qt',        when='+qt')
    depends_on('py-pyside', when='+qt', type=nolink)
```

We had a discussion about this in #1041. The ideal solution is to add multi-value variants since there should only be one backend at a time. Of course we could make a bunch of different variants for each option and raise warnings every time someone tries to do multiple of them, but then we'll just get rid of all of that hard work once we switch to multi-value variants.

> ...plus maybe a bit of change on the filtering in the backends file at the end of the package?

This is a lot harder than it sounds. I tried to do this once and found myself quickly lost. Good luck.

> In the meantime, can we use type='build' for them? 

Yep, or just comment them out if you want. That's what I've done in other packages.

**Comment by citibeth** [comment]
On Thu, Oct 6, 2016 at 3:39 PM, Adam J. Stewart <notifications@github.com>
wrote:

> So, I would suggest the following variants:
>
> Works for me. Spack has texinfo and texlive packages, which I believe are
> the equivalent of MikTeX for Linux.
>
OK.  Hopefully I won't be stuck building them.  (This falls WAY outside of
my needs).


> Why not just:
>
>     depends_on('tk@8.3:',   when='+tk')  # not 8.6.0 or 8.6.1
>     depends_on('qt',        when='+qt')
>     depends_on('py-pyside', when='+qt', type=nolink)
>
> We had a discussion about this in #1041
> <https://github.com/LLNL/spack/pull/1041>. The ideal solution is to add
> multi-value variants since there should only be one backend at a time. Of
> course we could make a bunch of different variants for each option and
> raise warnings every time someone tries to do multiple of them, but then
> we'll just get rid of all of that hard work once we switch to multi-value
> variants.
>

No... Matplotlib can be built to support multiple backends.  Although there
might not be a big NEED to build Matplotlib with simultaneous Tk and Qt
support, it is certainly capable of doing that.  For this case, boolean
variants are exactly what we want.

> ...plus maybe a bit of change on the filtering in the backends file at the
> end of the package?
>
> This is a lot harder than it sounds. I tried to do this once and found
> myself quickly lost. Good luck.
>

My preference is to not monkey with it.  Setting up the preferred backend
inside your $HOME directory is a standard part of starting with
Matplotlib.  Maybe add a comment to the Package explaining that...

> In the meantime, can we use type='build' for them?
>
> Yep, or just comment them out if you want. That's what I've done in other
> packages.
>
>
If they're commented out, how will the tests run?

-- Elizabeth

**Comment by citibeth** [comment]
Lots of interesting discussion on this thread, but none of it has addressed my core issue: that Basemap isn't working.  A little more research, I'm pretty sure this is the issue:

https://github.com/Homebrew/homebrew-python/issues/112

> The problem is that basemap is installed in your user site-packages and matplotlib is installed into another prefix, so the mpl_toolkits package is being shadowed.

Inspection of the Spack install directories confirms this theory.  Apparently, Homebrew solves this problem by packaging the two together.  Not sure why it used to work for me and now it doesn't...

**Comment by citibeth** [comment]
OK, a bit more digging... this is yet another implicit assumption that two packages are installed in the same source tree --- except now they are not.

I tried using Spack activate.  Didn't work.  I tried using Spack filesystem views to merge the two install trees.  That didn't work either.  Further investigation revealed that matplotlib-1.5.1 installs as an "egg", whereas basemap doesn't.  That's why even when the two trees are merged, they remain in separate directories.

My best guess at this point is something changed in matplotlib-1.5 that broke this in Spack.  I'm reverting to matplotlib-1.4.3, we'll see how that works.

**Comment by adamjstewart** [comment]
Is there any chance that this was broken by #1894, which told Spack to no longer add dependencies of dependencies to RPATH (and possibly other environment variables like PYTHONPATH)? Can you try again now that #1954 has been merged to revert this?

**Comment by citibeth** [comment]
No, it's because Matplotlib and Basemap assume they're installed in the
same tree.  If they're not, they do things that are known to be illegal in
Python.  It turns out, an older build of mine from August had the same
problem (but apparently I never tried using Basemap in August so I didn't
notice).

I'm on to this issue...

**Comment by adamjstewart** [comment]
@citibeth I believe you solved this problem for yourself in #1965? For any issues involving other versions of Python that weren't covered by your patch, let's discuss those in #3813.
